feat(scan): add expect method to wait using a custom predicate

ostridm opened the pull request:
closes #55

ostridm committed:
feat(scan): add expect method to wait using a custom predicate

closes #55

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +96,34 @@ public async Task Stop()
+      throw new ArgumentNullException(nameof(expectation));
+    }
+
+    using var cancellationTokenSource = new CancellationTokenSource();
+
+    cancellationTokenSource.CancelAfter(this._options.Timeout);
```
```suggestion
    using var cancellationTokenSource = new CancellationTokenSource(this._options.Timeout);
```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +96,34 @@ public async Task Stop()
+    await foreach (var _ in Status(CancellationToken.None).ConfigureAwait(false))
+    {
+      var preventFurtherPolling =
+        await expectation(this).ConfigureAwait(false) || Done || cancellationToken.IsCancellationRequested;
```
The 
```suggestion
        await predicate(this).ConfigureAwait(false) || Done || cancellationToken.IsCancellationRequested;
```

derevnjuk replied:
The predicate must be safe, you should wrap the function into try/catch

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -346,4 +327,92 @@ public async Task DisposeAsync_DeleteScanThrows_DoesNothing()
+
+  [Theory]
+  [MemberData(nameof(ActiveStatuses))]
+  public async Task Expect_GivenPredicateInvokingStatus_Returns(ScanStatus scanStatus)
```
GivenPredicateInvokingStatus?

ostridm committed:
feat(scan): add expect method to wait using a custom predicate

adjust according to codereview
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1040869088
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1040870415
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1040870415
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1040882516

closes #55

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +96,45 @@ public async Task Stop()
+    await Expect(predicate, cancellationTokenSource.Token).ConfigureAwait(false);
+  }
+
+  private async Task<bool> SatisfyPredicate(Func<Scan, Task<bool>> predicate)
```
This is quite a weird name. To say if your argument is satisfied predicated, you have to apply it to the argument first.
```suggestion
  private async Task<bool> ApplyPredicate(Func<Scan, Task<bool>> predicate)
```

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +96,45 @@ public async Task Stop()
     yield return _scanState;
   }
 
+  public async Task Expect(Func<Scan, Task<bool>> predicate)
```
We have already discussed that the user should be able to abort any long-running transient operation. To combine multiple sources, use `CancellationTokenSource.CreateLinkedTokenSource` 

```suggestion
  public async Task Expect(Func<Scan, Task<bool>> predicate,  CancellationToken? cancellationToken = default)
```

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -346,4 +327,115 @@ public async Task DisposeAsync_DeleteScanThrows_DoesNothing()
+      IssuesBySeverity = new[] { new IssueGroup(1, Severity.High) }
+    };
+
+    async Task<bool> Predicate(Scan scan)
```
Use lambda expression

derevnjuk replied:
So, why don't add 3rd level of predicate?) You should use NSubstitute to operate, verify, and capture arguments. We can leave this as is, but you should put a TODO comment

derevnjuk commented on test/SecTester.Scan.Tests/ScanTests.cs:
```diff
@@ -346,4 +327,115 @@ public async Task DisposeAsync_DeleteScanThrows_DoesNothing()
+
+      await _predicate(scan);
+
+      return status.IssuesBySeverity.Any(x => x.Type == Severity.High);
```
It may lead to npe

ostridm committed:
feat(scan): add expect method to wait using a custom predicate

adjust according to codereview
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1041869631
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1041875537
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1041876145
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1041873990
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1041878000

closes #55

derevnjuk commented on src/SecTester.Scan/Scan.cs:
```diff
@@ -96,6 +96,47 @@ public async Task Stop()
+
+    using var timeoutCancellationTokenSource = new CancellationTokenSource(this._options.Timeout);
+    using var linkedCancellationTokenSource =
+      CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCancellationTokenSource.Token);
```
```suggestion
    using var cancellationTokenSource =
      CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
    cancellationTokenSource.CancelAfter(_options.Timeout)
```

derevnjuk reviewed (changes requested):
Almost there, just a single remark

ostridm committed:
feat(scan): add expect method to wait using a custom predicate

adjust according to codereview
https://github.com/NeuraLegion/sectester-net/pull/86#discussion_r1042112956

closes #55
